---
title: "PAWS_mQTL"
author: "Rachel D Edgar"
date: "10/6/2016"
output: html_document
---
Correlation of methylation and SNPs
========================================================
### correlation with methylation


```{r}
setwd("~/bootcamp/mQTL")

library(reshape)
library(ggplot2)
library(RColorBrewer)
library(lme4)
library(gridExtra)
```


## SNPs
```{r}
load("~/bootcamp/PsychChip/PAWS_snps_filtered.RData")

      #for script testing
      all_snps<-unique(PAWS_snps_filtered$SNP_Name)
      SNPS<-as.character(all_snps[sample(1:length(all_snps), 10)])
      PAWS_snps_filtered<-PAWS_snps_filtered[which(PAWS_snps_filtered$SNP_Name%in%SNPS),]
      
      
      
## recode to 0,1,2 (BB=0, AB=1, AA=2)
Genotype<-sapply(1:nrow(PAWS_snps_filtered), function(x) {
  if(PAWS_snps_filtered$Allele1[x]=="B"){
    if(PAWS_snps_filtered$Allele2[x]=="B"){0}else{
      if(PAWS_snps_filtered$Allele2[x]=="A"){1}else{NA}}}else{
      
      if(PAWS_snps_filtered$Allele1[x]=="A"){
        if(PAWS_snps_filtered$Allele2[x]=="A"){2}else{
          if(PAWS_snps_filtered$Allele2[x]=="B"){1}else{NA}}}else{NA}}      
    })


PAWS_snps_filtered$Genotype<-Genotype
PAWS_snps_filtered_mini<-PAWS_snps_filtered[, c(1,2,9)]
PAWS_snps_filtered_cast<-cast(PAWS_snps_filtered_mini, SNP_Name~Sample_ID) #33039   193
PAWS_snps_filtered_cast<-as.data.frame(PAWS_snps_filtered_cast)

save(PAWS_snps_filtered_cast, file="PAWS_SNPs_Genotypes.RData")
```


#Heterozygosity Psych SNPs
```{r}
Heterozygosity_levels<-lapply(1:nrow(PAWS_snps_filtered_cast), function(x) {
  table(unlist(PAWS_snps_filtered_cast[x,2:ncol(PAWS_snps_filtered_cast)]))
  })
hets<-sapply(1:length(Heterozygosity_levels), function(x) length(Heterozygosity_levels[[x]]))#12884 no varibility in PAWS 10924 Hets and Homos 9231 hets and 2 Homos

table(hets)

PAWS_snps_variable<-PAWS_snps_filtered_cast[which(hets>1),]

## rename mixed sample
colnames(PAWS_snps_variable)[which(colnames(PAWS_snps_variable)=="10_9630002082_R10C02")]<-"186"
colnames(PAWS_snps_variable)[which(colnames(PAWS_snps_variable)=="10_9630002082_R08C02")]<-"10"

save(PAWS_snps_variable, file="PAWS_snps_variable.RData")
```




#DNAm data
```{r}
# this data may change to the VMR CpGs only
load("~/PAWS/combat_PAWS_Beta_norep.RData")



load("~/PAWS/SNPCpG.RData")
SnpatCpG<-SNPCpG[which(SNPCpG$SNPCpG!=""),]
PAWS_Beta<-PAWS_Beta[which(!(rownames(PAWS_Beta)%in%rownames(SnpatCpG))),]
PAWS_Beta<-as.data.frame(PAWS_Beta)

#M value transformation
Mval<-function(beta) log2(beta/(1-beta))
PAWS_Mval = apply(PAWS_Beta, 1, Mval) # need mvalues for combat
PAWS_Mval = as.data.frame(PAWS_Mval)
PAWS_Mval = t(PAWS_Mval)

# Betas  409878    178
# SNPS 20155   178
```



#add annoation information with the fixed the chromosome coordiante
```{r}
Psychchip_manifest<-read.csv("~/bootcamp/PsychChip/Psych_array_annotation.csv", sep="\t") #558741/571078 have a TOP BOT id :)
Psychchip_manifest_simple<-Psychchip_manifest[,c(1:3)]

Psychchip_manifest_simple<-Psychchip_manifest_simple[which(Psychchip_manifest_simple$Name%in%PAWS_snps_variable$SNP_Name),]
```


## which CpGs are within 1Mb of a SNP
```{r}
load("~/Price_annotation.RData")
annotation$CpG<-rownames(annotation)
annotation_simple<-annotation[,c("CpG","MAPINFO","CHR")]

PAWS_CpGs<-as.data.frame(annotation[which(annotation$CpG%in%rownames(PAWS_Beta)),])
Psychchip_manifest_simple$Name<-as.character(Psychchip_manifest_simple$Name)

## Only want CpGs on chr in PsychChip
Psychchip_manifest_simple$Chr<-as.character(Psychchip_manifest_simple$Chr)
annotation_simple$CHR<-as.character(annotation_simple$CHR)
annotation_simple<-annotation_simple[which(annotation_simple$CHR%in%unique(Psychchip_manifest_simple$Chr)),]
Psychchip_manifest_simple$Chr<-as.factor(Psychchip_manifest_simple$Chr)
annotation_simple$CHR<-as.factor(annotation_simple$CHR)

# split by chr
Psychchip_manifest_simple_split<-split(Psychchip_manifest_simple, Psychchip_manifest_simple$Chr)
annotation_simple_split<-split(annotation_simple, annotation_simple$CHR)

sapply(1:length(Psychchip_manifest_simple_split), function(x) Psychchip_manifest_simple_split[[x]]$Chr[1])
sapply(1:length(annotation_simple_split), function(x) annotation_simple_split[[x]]$CHR[1])

CpG_SNP_realtions<-lapply(1:length(annotation_simple_split), function(chr){
  chr_relations<-lapply(1:nrow(Psychchip_manifest_simple_split[[chr]]), function(x) {
    coor<-Psychchip_manifest_simple_split[[chr]][x,"MapInfo"]
    CpGs<-annotation_simple_split[[chr]][which(annotation_simple_split[[chr]]$MAPINFO>(coor-1000000) & annotation_simple_split[[chr]]$MAPINFO<(coor+1000000)),]
    if(nrow(CpGs)>0){
      data.frame(SNP=Psychchip_manifest_simple_split[[chr]][x,"Name"], CpGs=rownames(CpGs), distance=(coor-CpGs$MAPINFO))}
  })
  do.call(rbind, chr_relations)})

CpG_SNP_realtions<-do.call(rbind, CpG_SNP_realtions) 
```






## mQTLs
```{r}
load("~/bootcamp/PsychChip/PAWS_meta_sentrix_genetic_clusters.RData")
meta<-meta[which(meta$Meth_ID%in%colnames(PAWS_Beta)),]
meta<-meta[match(colnames(PAWS_Beta), meta$Meth_ID),]

## match
PAWS_snps_variable
CpGs_methylation
meta

mQTL_snp_cpg<-lapply(1:length(unique(CpG_SNP_realtions$SNP)), function(snp){
  snp=unique(CpG_SNP_realtions$SNP)[snp]
  snp_genotypes<-PAWS_snps_variable[which(rownames(PAWS_snps_variable)==snp),]
  CpGs<-CpG_SNP_realtions[which(CpG_SNP_realtions$SNP==snp),]
  CpGs_methylation<-PAWS_Mval[which(rownames(PAWS_Mval)%in%CpGs$CpGs),]
  
  CpGs$mQTL_pval<-sapply(1:nrow(CpGs), function(cpg){
    lm(CpGs_methylation~age+sex+geno1+geno2+smoke+snp_genotypes)$pval})})


mQTL_snp_cpg_unlist<-unlist(mQTL_snp_cpg)
CpG_SNP_realtions$mQTL_pval<-mQTL_snp_cpg_unlist

save(CpG_SNP_realtions, file="CpG_SNP_realtions.RData") 
```

